### ISSUE TITLE
Allow passing WebSocket implementation to WebsocketTransport?

### ISSUE THREAD

**Issue by achingbrain** [improbable-eng/grpc-web#796]
Would there be any interest from the maintainers in accepting a PR that allows you to pass a `WebSocket` implementation to `WebsocketTransport`?

It'd use the implementation if passed instead of trying to [instantiate one](https://github.com/improbable-eng/grpc-web/blob/master/client/grpc-web/src/transports/websocket/websocket.ts#L56) from the global scope.

That way you could pass in the [ws](https://www.npmjs.com/package/ws) module which would allow use of the `WebsocketTransport` in node.js.

**Comment by johanbrandhorst** [comment]
I think we try to discourage users from using the grpc-web library with NodeJS, where there is a perfectly good official gRPC client. Is there a compelling argument that we should allow this, which may encourage its use?

**Comment by achingbrain** [comment]
Bidirectional streaming is key to my use case and unless I missed something (entirely possible), the only way to do it with gRPC in the browser right now is to use websockets.

From what I can see the official gRPC-web client [doesn't support this](https://github.com/grpc/grpc-web/issues/24) and the APIs between that and `improbable-eng/grpc-web` are different which will mean two implementations will be necessary - I'd really rather not have to maintain both.

From my hacking the only thing stopping use of the `WebsocketTransport` in node.js is it trying to instantiate a `WebSocket` instance from the global context.

**Comment by johanbrandhorst** [comment]
You're correct that you can't get bi-directional gRPC in the browser without our websocket transport - but I'm not sure I understand the issue here, don't we already support NodeJS websocket servers? @MarcusLongmuir will know better than me how this works on the server side.

**Comment by MarcusLongmuir** [comment]
I think I'm lacking some understanding of the issue too. It seems that there's some likely contradictory points about trying to use gRPC from the browser, but then the issue is about trying to use a NodeJS module - implying the gRPC usage is actually within NodeJS rather than the browser.

This isn't strictly contradictory if you're doing some form server side rendering, but that information would be useful too.

If it weren't for the mention of usage in a browser then I'd be suggesting that you don't use gRPC-Web and instead just use regular gRPC from NodeJS using the official client.

**Comment by achingbrain** [comment]
Sorry for not being clear - I'm writing isomorphic client code that requires bidirectional streaming that is destined to run in both node and the browser talking to a remote gRPC server and trying to do it from the same codebase without having two separate client implementations.

`improbable-eng/grpc-web` with the `WebsocketTransport` on the client and a websocket->gRPC bridge on the server seems like a good fit for this, it just needs the ability to pass a `WebSocket` implementation as config somehow to let it run under node.
